### PR TITLE
Async cellular

### PR DESCRIPTION
**Pull request by jarvte**

### CODE REVIEW

File: features/cellular/framework/AT/AT_CellularDevice.cpp [hunk 1]
```diff
@@ -108,6 +108,8 @@ CellularNetwork *AT_CellularDevice::open_network(FileHandle *fh)
             _network = new AT_CellularNetwork(*atHandler);
             if (!_network) {
                 release_at_handler(atHandler);
+            } else {
+                // TODO: set urc's if sync mode
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularDevice.cpp]: TODO?

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 2]
```diff
@@ -28,28 +29,43 @@ using namespace mbed;
...
 AT_CellularNetwork::AT_CellularNetwork(ATHandler &atHandler) : AT_CellularBase(atHandler),
     _stack(NULL), _apn(NULL), _uname(NULL), _pwd(NULL), _ip_stack_type_requested(DEFAULT_STACK), _ip_stack_type(DEFAULT_STACK), _cid(-1),
-    _connection_status_cb(NULL), _op_act(operator_t::RAT_UNKNOWN), _authentication_type(CHAP), _last_reg_type(C_REG),
-    _connect_status(NSAPI_STATUS_DISCONNECTED), _new_context_set(false)
+    _connection_status_cb(NULL), _op_act(RAT_UNKNOWN), _authentication_type(CHAP), _last_reg_type(C_REG),
+    _connect_status(NSAPI_STATUS_DISCONNECTED), _new_context_set(false), _reg_status(NotRegistered)
 {
-
-    _at.set_urc_handler("NO CARRIER", callback(this, &AT_CellularNetwork::urc_no_carrier));
+    _urc_funcs[C_EREG] = callback(this, &AT_CellularNetwork::urc_cereg);
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: These could be done in init() as well..?

File: features/cellular/framework/API/CellularNetwork.h [hunk 3]
```diff
@@ -201,13 +200,38 @@ class CellularNetwork : public NetworkInterface
...
+        NWModeAutomatic = 0,    // automatic registering
+        NWModeManual,           // manual registering with plmn
+        NWModeDeRegister,       // deregister from network
+        NWModeSetOnly,          // set only <format> (for read command +COPS?), do not attempt registration/deregistration
+        NWModeManualAutomatic   // if manual fails, fallback to automatic
+    };
+
+
+    /** Does all the needed initializations that can fail
+     *
```
Comment (AriParkkila) [inline features/cellular/framework/API/CellularNetwork.h]: Make a remark that this must be called immediately after constructor.

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 4]
```diff
@@ -73,6 +89,47 @@ void AT_CellularNetwork::urc_no_carrier()
...
 
+void AT_CellularNetwork::read_reg_params_and_compare(RegistrationType type)
+{
+    RegistrationStatus reg_status = NotRegistered;
+    int lac = -1, cell_id = -1, act = -1;
+
+    read_reg_params(type, reg_status, lac, cell_id, act);
+
+    if (_at.get_last_error() == NSAPI_ERROR_OK) {
+        tr_debug("stat: %d, lac: %d, cellID: %d, act: %d", reg_status, lac, cell_id, act);
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: cb could be checked only once, like... if (_at.get_last_error() == NSAPI_ERROR_OK && _connection_status_cb)...

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 5]
```diff
@@ -28,28 +29,43 @@ using namespace mbed;
...
 AT_CellularNetwork::~AT_CellularNetwork()
 {
     free_credentials();
 }
 
+nsapi_error_t AT_CellularNetwork::init()
+{
+    for (int type = 0; type < CellularNetwork::C_MAX; type++) {
+        if (has_registration((RegistrationType)type)) {
+            if (_at.set_urc_handler(at_reg[type].urc_prefix, _urc_funcs[type])) {
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: could be easier to read with != NSAPI_ERROR_OK

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 6]
```diff
@@ -512,37 +566,48 @@ nsapi_ip_stack_t AT_CellularNetwork::string_to_stack_type(const char* pdp_type)
...
-                _at.write_string("=0", false);
-                _at.cmd_stop();
-            }
-
-            _at.resp_start();
-            _at.resp_stop();
+    int index = (int)type;
+    MBED_ASSERT(index >= 0 && index < C_MAX);
+
+    if (has_registration(type)) {
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: flat code:  if (!has_registration(type)) return...

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 7]
```diff
@@ -512,37 +566,48 @@ nsapi_ip_stack_t AT_CellularNetwork::string_to_stack_type(const char* pdp_type)
...
-            } else {
-                _at.cmd_start(at_reg[i].cmd);
-                _at.write_string("=0", false);
-                _at.cmd_stop();
-            }
-
-            _at.resp_start();
-            _at.resp_stop();
+    int index = (int)type;
+    MBED_ASSERT(index >= 0 && index < C_MAX);
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: MBED_ASSERT may be disabled in some builds, which could give warning: "unused variable index".

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 8]
```diff
@@ -512,37 +566,48 @@ nsapi_ip_stack_t AT_CellularNetwork::string_to_stack_type(const char* pdp_type)
...
 
-nsapi_error_t AT_CellularNetwork::set_registration(const char *plmn)
+nsapi_error_t AT_CellularNetwork::get_network_registering_mode(NWRegisteringMode& mode)
 {
     _at.lock();
+    _at.cmd_start("AT+COPS?");
+    _at.cmd_stop();
+    _at.resp_start("+COPS:");
+    int tmpmode = (NWRegisteringMode)_at.read_int();
+    _at.resp_stop();
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: Really need here tmpmode variable and casting..?

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 9]
```diff
@@ -607,38 +645,58 @@ nsapi_error_t AT_CellularNetwork::get_registration_status(RegistrationType type,
...
+        _last_reg_type = type;
+    }
 }
 
-nsapi_error_t AT_CellularNetwork::get_cell_id(int &cell_id)
+nsapi_error_t AT_CellularNetwork::get_registration_status(RegistrationType type, RegistrationStatus &status)
 {
-    RegistrationStatus tmp;
+    int i = (int)type;
+    MBED_ASSERT(i >= 0 && i < C_MAX);
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: same rant as above

File: features/cellular/framework/AT/AT_CellularNetwork.cpp [hunk 10]
```diff
@@ -607,38 +645,58 @@ nsapi_error_t AT_CellularNetwork::get_registration_status(RegistrationType type,
...
+
+    _at.lock();
+
+    const char *rsp[] = { "+CEREG:", "+CGREG:", "+CREG:"};
+    _at.cmd_start(at_reg[i].cmd);
+    _at.write_string("?", false);
+    _at.cmd_stop();
+    _at.resp_start(rsp[i]);
+
+    _at.read_int(); // ignore urc mode subparam
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularNetwork.cpp]: (void) casting

File: features/cellular/framework/AT/AT_CellularPower.cpp [hunk 11]
```diff
@@ -43,6 +43,8 @@ nsapi_error_t AT_CellularPower::off()
     return NSAPI_ERROR_UNSUPPORTED;
 }
 
+// wiat filehandle irq
```
Comment (AriParkkila) [inline features/cellular/framework/AT/AT_CellularPower.cpp]: ???

File: features/netsocket/nsapi_types.h [hunk 12]
```diff
@@ -73,13 +73,14 @@ enum nsapi_error {
...
- * 
+ *
  *  Event callbacks are accompanied with an event-dependent parameter passed as an intptr_t.
  *
  *  @enum nsapi_event
  */
  typedef enum nsapi_event {
-    NSAPI_EVENT_CONNECTION_STATUS_CHANGE = 0 /*!< network connection status has changed, the parameter = new status (nsapi_connection_status_t) */
+    NSAPI_EVENT_CONNECTION_STATUS_CHANGE = 0, /*!< network connection status has changed, the parameter = new status (nsapi_connection_status_t) */
+    NSAPI_EVENT_CELLULAR_STATUS_CHANGE   = 1  /*!< cellular modem status has changed, the parameter = new status (cellular_connection_status_t) */
```
Comment (AriParkkila) [inline features/netsocket/nsapi_types.h]: Kevin also suggested that we could request "a chunk of number space like 0x1000-0x10ff from nsapi_event_t to give you a range of events, so each of them can still have a useful parameter".

File: features/cellular/framework/API/CellularNetwork.h [hunk 13]
```diff
@@ -271,10 +295,12 @@ class CellularNetwork : public NetworkInterface
...
      *
-     *  @param op_rat Radio access technology
-     *  @return       zero on success
+     *  @param rat  Radio access technology
+     *  @return     zero on success
      */
-    virtual nsapi_error_t set_access_technology(operator_t::RadioAccessTechnology op_rat) = 0;
+    virtual nsapi_error_t set_access_technology(RadioAccessTechnology rat) = 0;
+
+    virtual nsapi_error_t get_access_technology(RadioAccessTechnology& rat) = 0;
```
[the remark this answers is not in the document]
Comment (jarvte) [reply features/cellular/framework/API/CellularNetwork.h]: check

File: features/cellular/framework/AT/ATHandler.cpp [hunk 14]
```diff
@@ -150,21 +150,45 @@ void ATHandler::set_file_handle(FileHandle *fh)
     _fileHandle = fh;
 }
 
-void ATHandler::set_urc_handler(const char *prefix, mbed::Callback<void()> callback)
+nsapi_error_t ATHandler::set_urc_handler(const char *prefix, mbed::Callback<void()> callback)
 {
+    if (check_urc_existance(prefix, callback)) {
+        tr_warn("URC already added with prefix: %s", prefix);
```
[the remark this answers is not in the document]
Comment (jarvte) [reply features/cellular/framework/AT/ATHandler.cpp]: check

File: features/cellular/framework/AT/AT_CellularDevice.cpp (hunk 1 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularDevice.cpp]: removed

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 2 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: check

File: features/cellular/framework/API/CellularNetwork.h (hunk 3 above)
Comment (jarvte) [reply features/cellular/framework/API/CellularNetwork.h]: check

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 4 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: good catch, fixed

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 5 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: true, fixed

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 6 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: fixed

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 8 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: casting yes, tmpmode no...

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 10 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: fixed

File: features/cellular/framework/AT/AT_CellularPower.cpp (hunk 11 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularPower.cpp]: removed :)

File: features/netsocket/nsapi_types.h (hunk 12 above)
Comment (jarvte) [reply features/netsocket/nsapi_types.h]: NSAPI_EVENT_CONNECTION_STATUS_CHANGE is already used in this way so we should this in same way for consistence.

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 9 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: i and type are used so no warnings

File: features/cellular/framework/AT/AT_CellularNetwork.cpp (hunk 7 above)
Comment (jarvte) [reply features/cellular/framework/AT/AT_CellularNetwork.cpp]: index and type are used so no warnings

File: features/cellular/easy_cellular/CellularConnectionFSM.cpp [hunk 15]
```diff
@@ -268,223 +334,197 @@ nsapi_error_t CellularConnectionFSM::continue_to_state(CellularState state)
...
+    } else {
+        retry_state_or_fail();
+    }
+}
+
+void CellularConnectionFSM::state_registering()
+{
+    _cellularDevice->set_timeout(TIMEOUT_NETWORK);
+
+    if (is_registered()) {
```
Comment (AriParkkila) [inline features/cellular/easy_cellular/CellularConnectionFSM.cpp]: Could we trigger COPS=0 once and then rely on CREGs to report connection status due to COPS=0 will block during while AT commands will fail.

File: features/cellular/easy_cellular/CellularConnectionFSM.cpp [hunk 16]
```diff
@@ -90,18 +111,16 @@ nsapi_error_t CellularConnectionFSM::init()
...
-bool CellularConnectionFSM::open_power(FileHandle *fh)
+bool CellularConnectionFSM::power_on()
 {
-    if (!_power) {
-        _power = _cellularDevice->open_power(fh);
-        if (!_power) {
-            return false;
-        }
-    }
+    MBED_ASSERT(1);
```
[the remark this answers is not in the document]
Comment (jarvte) [reply features/cellular/easy_cellular/CellularConnectionFSM.cpp]: removed...

[the remark this answers is not in the document]
Comment (jarvte) [reply features/cellular/easy_cellular/CellularConnectionFSM.cpp]: locally...

**Pull request closed without merging**